lib/eblox/data/taxonomy.ex

```diff
@@ -0,0 +1,73 @@
+defmodule Eblox.Data.Taxonomy do
+  @moduledoc false
+
+  use Supervisor
+
+  @type post_id() :: term()
+  @type registry_option() :: Registry.start_option()
+
+  @callback registry_options([registry_option()]) :: [registry_option()]
+  @callback on_add(module(), post_id(), term()) :: :ok | :error
+  @callback on_remove(module(), post_id()) :: :ok
+
+  defmodule Meta do
+    @moduledoc false
+
+    use Agent
+
+    alias Eblox.Data.Taxonomy.Meta
+
+    @type t :: %{
+      __struct__: Meta,
+      impl: module(),
+    }
+    @enforce_keys ~w|impl|a
+    defstruct ~w|impl|a
+
+    def start_link([{:impl, impl} | opts]) do
+      Agent.start_link(fn -> %Meta{impl: impl} end, opts)
+    end
+
+    def impl(name) do
+      Agent.get(name, &(&1 |> Map.get(:impl)))
+    end
+  end
+
+  def start_link(opts), do:
+    Supervisor.start_link(__MODULE__, opts)
+
+  def init(opts \\ []) do
+    {impl, opts} = Keyword.pop!(opts, :impl)
+    {name, _opts} = Keyword.pop(opts, :name, impl)
+
+    children = [
+      {Meta, impl: impl, name: meta_name(name)},
+      {Registry, impl.registry_options(name: reg_name(name))}
+      # {PubSub, name: pubsub_name(name)}
+    ]
+
+    Supervisor.init(children, name: sup_name(name), strategy: :one_for_one)
+  end
+
+  def add(name, post_id, value \\ nil) do
```

Comment:
What would be the use case of adding a post without value? Why do we need this default `\\ nil`?

Reply:
I believe it depends on a taxonomy implementation. But since we do polymorphism here I think you're right - it should follow the uniform contract.

Reply:
> we do polymorphism here

There is no issue to have default params in the behaviour implementation, just define both callbacks

```elixir
  @callback on_add(module(), post_id()) :: :ok | :error
  @callback on_add(module(), post_id(), term()) :: :ok | :error

  @optional_callbacks on_add: 2
```
_But_ there is a great rule: don’t do things just in case. When the case happens, we’ll easily amend this.

Reply:
I decided we don't need the `value` parameter for now. The value is supposed to be derived from `post_id` until additional input is required in one of the future taxonomy implementations.